src/textual/widgets/_input.py

```diff
@@ -32,7 +32,7 @@
 
 _RESTRICT_TYPES = {
     "integer": r"[-+]?\d*",
-    "number": r"[-+]?\d*\.?\d*[eE]?-?\d*",
+    "number": r"[-+]?\d*\.?\d*[eE]?[-+]?\d*",
```

Comment:
Looks like I've missed the boat on this one now, but while out running I realised there's one more thing should probably happen here: at the very least the decimal point in the above should use `locale.localeconv()["decimal_point"]` rather than hard-code `.`.

I'll add an issue.